Check that PC (or PCC.cursor for CHERI-RISCV) was updated before reading it

arichardson opened the pull request.

arichardson committed:
Check that PC (or PCC.cursor for CHERI-RISCV) was updated before reading it

This is a rather invasive change but should avoid lots of subtle bugs in
the future. It is only on when --enable-debug-tcg is passed to configure so
does not have any impact on run-time performance.

arichardson committed:
[CHERI-RISCV] Don't save PC for all CHERI instructions

This is not necessary since we can lazily compute the current value when an
exception is triggered.

nwf commented on target/cheri-common/cheri-helper-utils.h:
```diff
@@ -41,6 +41,54 @@
+
+static inline const cap_register_t *_cheri_get_pcc_unchecked(CPUArchState *env);
+
+// Returns a recebt PCC value. This means that the cursor field may be out of
```
rece*n*t

nwf commented on target/cheri-common/cheri-bounds-stats.h:
```diff
@@ -93,7 +93,9 @@ struct oob_stats_info {
-static inline int64_t _howmuch_out_of_bounds(CPUArchState *env, const cap_register_t* cr, const char* name)
+static inline int64_t _howmuch_out_of_bounds(CPUArchState *env,
+                                             const cap_register_t *cr,
+                                             const char *name, uintptr_t retpc)
```
It may be worth a comment that, despite taking both `CPUArchState` and `retpc`, this function does not throw CPU exceptions (that is, `retpc` is there purely for diagnostic reporting)?

nwf commented on target/cheri-common/cheri-helper-utils.h:
```diff
@@ -41,6 +41,54 @@
+// by calling cpu_restore_state(). However, it is preferable if the translator
+// code writes the updated pcc.cursor before calling the helper if this is the
+// common case (for example, it should not be saved if the current pc value is
+// only needed if a trap is raised).
```
For my own edification, I assume the translator does PCC bounds checks when it generates TCG opcodes?

arichardson replied:
Currently the PCC bounds checks are done in ccheck_pc which is a helper that is called for every instruction. One of the main motivating reasons for this change it to be able to move the bounds checks into the translator code.

nwf commented on target/cheri-common/cheri-helper-utils.h:
```diff
@@ -51,15 +99,17 @@ static inline void derive_cap_from_pcc(CPUArchState *env, uint32_t cd,
 #endif
-    const cap_register_t *pccp = cheri_get_pcc(env);
+    // Note: we can use a PCC value with an outdated cursor since we are writing
+    // a new address.
```
And the address we're writing is guaranteed to be in bounds?

arichardson replied:
No, but in that case we get an oob/untagged value (e.g. auipcc, cgetpccincoffset). I've clarified the comment.

arichardson replied:
Turns out cgetpccincoffset wasn't saving PC. Thanks for reminding me to check!

nwf commented on target/mips/cpu.h:
```diff
@@ -1476,9 +1479,36 @@ static inline bool in_kernel_mode(CPUMIPSState *env) {
+{
+    state->PC = pc_addr;
+#ifdef TARGET_CHERI
+    // Note: no representability check, those are done on jump.
```
Is that to say they've already been done by the time we get here?

arichardson replied:
Not a particularly good comment. Generally this helper should only be called with in-bounds values so we don't need a check. In cases where it may be out of bounds we will (or at least should) be doing a check later.

arichardson committed:
Move some CHERI helpers to a separate header

Add a new cpu_cheri.h header that can be included from cpu.h once
CPUArchState has been defined. This works around some cyclic dependencies.

nwf commented on target/mips/helper.c:
```diff
@@ -585,7 +585,6 @@ static void raise_mmu_exception(CPUMIPSState *env, target_ulong address,
         /* TLB capability store bit was set, blocking capability store. */
         cpu_mips_store_capcause(env, reg, CapEx_TLBNoStoreCap);
-        env->active_tc.PC = cap_get_cursor(&env->active_tc.PCC);
         exception = EXCP_C2E;
```
Above, add a patch hunk to change `exception` from an `int` to a `MipsExcp`?

nwf reviewed (approved):
Some very minor nits, but it otherwise LGTM.

arichardson committed:
Save PC before calling the dumpstate helpers

Found using the new assertions

arichardson committed:
[CHERI-MIPS] Use PCC.cursor instead of PC

There is no need to keep both around as it is easy for them to become
out of sync. Use the RISCV approach of only using PCC for TARGET_CHERI

arichardson committed:
Improve checks and comment in PCC update helper

arichardson committed:
Save PCC.cursor before calling cgetpccincoffset helper

arichardson committed:
Use MipsExcp enum

Results in an improved debugging experience.

arichardson committed:
A bit more detail for "stale pcc is okay" comment

arichardson committed:
Explain that _howmuch_out_of_bounds does not trap

Using the comment suggested by @nwf in https://github.com/CTSRD-CHERI/qemu/pull/84

arichardson committed:
[UPSTREAM?] Fix missing gen_save_pc() before eret

Found with the new https://github.com/CTSRD-CHERI/qemu/pull/84 checks. This
shouldn't matter in practice since the value is only used for a debug log
message but printing he correct value is helpful.
